autorest/autorest_test.go

```diff
@@ -0,0 +1,79 @@
+package autorest
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/helpers"
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/iam"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2017-06-01/storage"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+var accountName = "testname01"
+
+func requestInspect() autorest.PrepareDecorator {
+	return func(p autorest.Preparer) autorest.Preparer {
+		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
+			log.Printf("Inspecting Request: %s %s\n", r.Method, r.URL)
+			return p.Prepare(r)
+		})
+	}
+}
+
+func respondInspect() autorest.RespondDecorator {
+	return func(r autorest.Responder) autorest.Responder {
+		return autorest.ResponderFunc(func(resp *http.Response) error {
+			log.Printf("Inspecting Response: %s for %s %s\n", resp.Status, resp.Request.Method, resp.Request.URL)
+			return r.Respond(resp)
+		})
+	}
+}
+
+// ExampleHookPipeline demonstrates how to add hooks to the request/response pipeline
+func ExampleHookPipeline() {
+	helpers.ParseArgs()
+	var logfile = "./output.log"
+	f, err := os.Create(logfile)
+	if err != nil {
+		log.Fatalf("failed to create file: %s", err)
+	}
+	defer os.Remove(logfile)
+	log.SetOutput(f)
+
+	token, _ := iam.GetResourceManagementToken(iam.OAuthGrantTypeServicePrincipal)
+	ac := storage.NewAccountsClient(helpers.SubscriptionID())
+	ac.Authorizer = autorest.NewBearerAuthorizer(token)
+
+	ac.Sender = autorest.CreateSender(
+		autorest.WithLogging(log.New(f, "example: ", log.LstdFlags)),
+	)
+	ac.RequestInspector = requestInspect()
+	ac.ResponseInspector = respondInspect()
+
+	nameAvailabilityResponse, err := ac.CheckNameAvailability(
+		storage.AccountCheckNameAvailabilityParameters{
+			Name: to.StringPtr(accountName),
+			Type: to.StringPtr("Microsoft.Storage/storageAccounts"),
+		},
+	)
+
+	if err != nil {
+		log.Fatalf("Error: %v", err)
```

Comment:
Using [`log.Fatalf`](https://godoc.org/log#Fatal) will bypass your defer statements. That would mean that your logFIle would get cleaned-up. There's a certain logic to that, but I'd rather be more explicit about those conditions rather than relying on defer semantics.

Reply:
For what its worth, the way I deal with this is combining a [`log.Printf`](https://godoc.org/log#Printf) call with a return statement.